### ISSUE TITLE
Struggling with high memory usage

### ISSUE THREAD

**Issue by rjpowers10** [OrchardCMS/OrchardCore#14117]
### Describe the bug
My scenario: We have a site built on Orchard Core 1.6. We have a pipeline that periodically deploys an instance to a test server and runs a suite of UI tests against it (the tests are using MSTest and Playwright, if it makes any difference). The site is running in IIS. The test suite runs for about 1 hour, and by the end of that run the w3wp process is consuming over 8.5 GB. `ASPNETCORE_ENVIRONMENT` is set to "Development".

Here's a quick look at the memory usage over time. This is the memory of the entire machine, not just OC, but looking in Task Manager, I see w3wp using over 8.5 GB.

![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/f8bf385f-b648-4b43-98c8-50f7b6cd0592)

### To Reproduce
I've been struggling for quite a while on pinpointing the root issue, with no success thus far. I've tried using the Visual Studio memory profiler but I've been struggling to make any sense of it.

I next tried a much simpler example. This next example is a plain old OC site (latest from the main branch) using the blog recipe. That's it. `ASPNETCORE_ENVIRONMENT` is set to "Development". I then wrote a small PowerShell script to ping the site over and over.

```
function Invoke-Path($path) {
    Invoke-WebRequest -Uri "https://localhost:8001/${path}"
}

$Paths = @(
    '', 
    'categories', 
    'tags', 
    'Contents/ContentItems/40rg5j34w43er1a5dwg0j14cnr', 
    'login'
)

while (1) {
    $Paths | ForEach-Object -Process {
        Invoke-Path $_
    }
}
``` 

I ran this script for a little while. Looking at the diagnostic tools in Visual Studio, I can see multiple garbage collector invocations (the yellow arrows), but the memory never seems to decrease. I was hoping to see a more jagged pattern in the memory usage.

![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/0694663e-1c41-43a8-93c7-7f3c7f45bbc9)

![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/ed470573-8395-4414-87bd-eb138f8f282c)

At this point I'm looking for any ideas or advice. Thanks.

**Comment by MikeAlhayek** [comment]
@rjpowers10 have you tried to disable cache from the general site settings? I am guessing there is a cache issue of maybe too many cached object in memory?

**Comment by MikeAlhayek** [comment]
Have you tried to disable all sites except for the default one? This may help trying to understand what is causing the high usage. If the default site has normal memory usage, try to enable one of your blog sites and see if you can pin point the site. Then maybe start disabling one feature at a time to see if you can pin point the problem to a single feature. This may give us more clues. 

Also, check your log file for any silent errors that would be causing memory consumption. 

Not sure if this will help, but something for now. Hopefully, @jtkech or @sebastienros have more ideas here.

**Comment by sebastienros** [comment]
How much memory is on the system in total. Take a memory snapshot between two jumps and check the diff. Run in Release, the fact that I see VS might mean it is a debug session.

**Comment by rjpowers10** [comment]
> How much memory is on the system in total. Take a memory snapshot between two jumps and check the diff. Run in Release, the fact that I see VS might mean it is a debug session.

My first example (my app built on OC) is a release build. My second example, which was a plain old OC site with the blog recipe, was in debug. The plain old OC site might just be a red herring, I just found it curious that memory didn't seem to go down after a GC collect. May not be relevant at all to my actual scenario.

The machine has 16GB of memory. SQL Server is also running which takes about 2GB itself. With my monitor I see the memory getting down to 0 and then as expected, the system starts paging and my tests start timing out.

I have a memory diff but I'm having trouble interpreting it. Wondering if anything jumps out to anyone.

Sorted by count diff
![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/af6081c9-ee55-4d3a-82e6-d16ac94b4524)

Sorted by size diff
![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/5c736e4f-2567-414e-adeb-f2221c2435f2)

Sorted by inclusive size diff
![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/5fb43abd-f123-47a2-bd00-c266216de374)

**Comment by rjpowers10** [comment]
> Have you tried to disable all sites except for the default one? This may help trying to understand what is causing the high usage. If the default site has normal memory usage, try to enable one of your blog sites and see if you can pin point the site. Then maybe start disabling one feature at a time to see if you can pin point the problem to a single feature. This may give us more clues.
> 
> Also, check your log file for any silent errors that would be causing memory consumption.
> 
> Not sure if this will help, but something for now. Hopefully, @jtkech or @sebastienros have more ideas here.

I'm going to try limiting my test suite. Maybe there is one test in particular that exemplifies the problem.

**Comment by sebastienros** [comment]
RouteEndpoints in the diff would be explained by a tenant being loaded.

Coming back to your original description, the issue is that your test is taking 8.5Gb, can you try to do the same thing with ASPNETCORE_ENVIRONMENT in Production, not "Development", just to see if that has an impact.

Can you take a memory dump when the app is close to the end to see what is still kept in memory?

**Comment by rjpowers10** [comment]
> RouteEndpoints in the diff would be explained by a tenant being loaded.
> 
> Coming back to your original description, the issue is that your test is taking 8.5Gb, can you try to do the same thing with ASPNETCORE_ENVIRONMENT in Production, not "Development", just to see if that has an impact.
> 
> Can you take a memory dump when the app is close to the end to see what is still kept in memory?

I ran the test suite again, this time with ASPNETCORE_ENVIRONMENT set to 'Production'. It didn't seem to make a difference. The tests finished in 1 hour, 15 minutes. This is what it looked like at the end of the test run.

![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/9b891869-d568-43be-9a7c-65aa282ce7c2)

This is the memory dump at the conclusion of the run.

Sorted by count
![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/03369eb0-a6a3-4d6a-9281-ed3fb15f3214)

Sorted by size
![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/87fdeb9d-8ca1-4bcd-8dbf-339cab1d90dd)

Sorted by inclusive size
![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/3cf5bd7c-3a48-4ae2-bfa8-b2592cd7dfe1)

**Comment by ShaneCourtrille** [comment]
Our memory issues are consistent with this as well.  These are our worst offenders in terms of memory usage from our latest performance testing while trying (and failing miserably) to run with 1000 tenants.

```
000071c5d1578080   449841     25195716 System.Int32[]
000071c5d15de7f0   101131     25791704 System.Byte[]
000071c5daff7f50   585044     28082112 System.Collections.Concurrent.ConcurrentDictionary`2+Node[[System.Type, System.Private.CoreLib],[System.Object[], System.Private.CoreLib]]
000071c5d7d8fbd0   728944     29157760 Microsoft.AspNetCore.Routing.RouteValueDictionary
000071c5d56e17c0   572454     32057424 System.Collections.Generic.Dictionary`2+Enumerator[[System.String, System.Private.CoreLib],[System.String, System.Private.CoreLib]]
000071c5daff8000   138733     37735376 System.Collections.Concurrent.ConcurrentDictionary`2+Node[[System.Type, System.Private.CoreLib],[System.Object[], System.Private.CoreLib]][]
000071c5d14ab110   499016     45688856 System.Object[]
000071c5d1e0d050    31347     47360136 System.Reflection.CustomAttributeRecord[]
000071c5d3fc6db8   963130     78553616 System.Collections.Generic.KeyValuePair`2[[System.String, System.Private.CoreLib],[System.Object, System.Private.CoreLib]][]
000071c5d157d2e0  1208078    108278276 System.String
Total 16274388 objects
```

**Comment by sebastienros** [comment]
Time to do another pass at it!

I will try with my own "script" but I'd like to be able to repro something that matches your experience. @rjpowers10 how many tenants are you loading and what features do they use? Ideally if you could do the same experiment and confirm it with the smallest possible list of features that would help. @ShaneCourtrille thanks for the confirmation. I'd love to achieve a density of 1000 on 8GB of RAM, we were there at the beginning, but obviously more features and complexity have been added since then.

**Comment by rjpowers10** [comment]
@sebastienros I'll try to help as much as I can though it may be difficult to repro my situation. We built an eCommerce platform on top of OC and wrote many custom features. Because I'm having trouble figuring out the offending code it's hard for me to say if it's an issue in my code or OC. My tests do a lot of the typical eCommerce things (sign in, register, view product, add to cart, checkout, order history, etc.).

I wonder if something is sticking around after each ViewResult so it's simply the fact that my tests are making a lot of requests. That's why I tried a simpler test with a plain old OC site and the blog recipe, but it's hard to say if that really replicated the issue.

List of OC features we have enabled:

1.  OrchardCore.Admin
2.  OrchardCore.Alias
3.  OrchardCore.Autoroute
4.  OrchardCore.BackgroundTasks
5.  OrchardCore.ContentFields
6.  OrchardCore.ContentPreview
7.  OrchardCore.Contents
8.  OrchardCore.ContentTypes
9.  OrchardCore.CustomSettings
10.  OrchardCore.Deployment
11.  OrchardCore.Diagnostics
12.  OrchardCore.DynamicCache
13.  OrchardCore.Email
14.  OrchardCore.Features
15.  OrchardCore.Flows
16.  OrchardCore.Forms
17.  OrchardCore.Google.Analytics
18.  OrchardCore.HomeRoute
19.  OrchardCore.Html
20.  OrchardCore.Https
21.  OrchardCore.Indexing
22.  OrchardCore.Layers
23.  OrchardCore.Liquid
24.  OrchardCore.Lists
25.  OrchardCore.Localization
26.  OrchardCore.Markdown
27.  OrchardCore.Media
28.  OrchardCore.Media.Slugify
29.  OrchardCore.Menu
30.  OrchardCore.Navigation
31.  OrchardCore.Placements
32.  OrchardCore.Recipes
33.  OrchardCore.Resources
34.  OrchardCore.Roles
35.  OrchardCore.Rules
36.  OrchardCore.Scripting
37.  OrchardCore.Search.Lucene
38.  OrchardCore.Search.Lucene.Worker
39.  OrchardCore.Settings
40.  OrchardCore.Shortcodes
41.  OrchardCore.Taxonomies
42.  OrchardCore.Themes
43.  OrchardCore.Title
44.  OrchardCore.Users
45.  OrchardCore.Users.ChangeEmail
46.  OrchardCore.Users.Registration
47.  OrchardCore.Users.ResetPassword
48.  OrchardCore.Widgets
49.  OrchardCore.Workflows        
50.  TheAdmin

**Comment by sebastienros** [comment]
@rjpowers10 no tenants?

**Comment by rjpowers10** [comment]
@sebastienros Oh sorry, we do have two tenants, one for humans and one for automated tests. All the tests are hitting the same tenant. The test are also running serially. The tenant for humans usually isn't doing much at this time though, since the tests run overnight.

**Comment by ShaneCourtrille** [comment]
@rjpowers10 Out of curiosity.. have you taken a look at your strings?  I'm doing some analysis and found an oddity where an estimated 80% of strings that are 46 bytes long are the value "mvc.1.0.view" being repeated over and over.  What's really painful is that these strings don't have a gcroot but are just sort of sitting around.

**Comment by rjpowers10** [comment]
> @rjpowers10 Out of curiosity.. have you taken a look at your strings? I'm doing some analysis and found an oddity where an estimated 80% of strings that are 46 bytes long are the value "mvc.1.0.view" being repeated over and over. What's really painful is that these strings (or at least the ones I've been able to sample since gcroot is such an expensive operation) don't have a gcroot but are just sort of sitting around.

There are a ton of strings in my memory dump, although only four instances of "mvc.1.0.view".

**Comment by wAsnk** [comment]
This might be a total red herring, but we have experienced increasing memory usage which can be reproduced by the following steps in vanilla OC, so I'd like to share my results. The main idea behind this was to create an idle tenant shutdown handler that would "shut down" tenants/free up resources for tenants that are not used for a predefined time period.

**Test controller for the repro:**
```csharp
using Microsoft.AspNetCore.Mvc;
using OrchardCore.Environment.Shell;

namespace OrchardCore.Cms.Web.Controllers;

[Route("test")]
public class TestController : Controller
{
    private readonly IShellHost _shellHost;
    private readonly ShellSettings _shellSettings;

    private static readonly HttpClient _httpClient = new();

    public TestController(
        IShellHost shellHost,
        ShellSettings shellSettings)
    {
        _shellHost = shellHost;
        _shellSettings = shellSettings;
    }

    public async Task<string> Index()
    {
        _httpClient.BaseAddress ??= new Uri(HttpContext.Request.Scheme + "://" + HttpContext.Request.Host);

        for (int i = 0; i < 1000; i++)
        {
            await _shellHost.ReleaseShellContextAsync(_shellSettings);

            await _httpClient.GetAsync(_shellSettings.RequestUrlPrefix, HttpContext.RequestAborted);
        }

        return "OK";
    }
}
```

**Repro steps:**
1. Use the OC repository.
2. Setup default tenant with Software as service recipe
3. Add a controller like the one above which uses the IShellHost.ReleaseShellContextAsync function which summary says: `Releases a shell so that a new one will be built for subsequent requests. Note: Can be used to free up resources after a given time of inactivity.` So we want to do this exactly.
4. Call /test URL, notice that after a while GC kicks in and the memory usage will maximize around ~500MB and never go above.
![image](https://github.com/OrchardCMS/OrchardCore/assets/19938236/37740d6b-0a8d-4a6e-b5bd-15687cdc6928)
5. Now enable ReCaptcha Users feature.
6. Call /test, notice how the memory usage goes up until the call is aborted (if not aborted it will use all memory), GC can't control the memory usage even after 10 minutes: 
![image](https://github.com/OrchardCMS/OrchardCore/assets/19938236/b5e9c011-74ab-4471-8298-d4a2a701c9f0)

Before and after the ReCaptcha Users feature memory snapshot results:
1. Count Diff
![image](https://github.com/OrchardCMS/OrchardCore/assets/19938236/4fd7ef0e-bcff-4584-a6d3-4c14696917cf)
2. Size Diff
![image](https://github.com/OrchardCMS/OrchardCore/assets/19938236/1ae6e4f3-3193-49b6-b430-7ad0c09c6009)
4. Inclusive Size Diff
![image](https://github.com/OrchardCMS/OrchardCore/assets/19938236/52056f9a-0393-418e-8659-d02d59b0aab4)
5. Insights
![image](https://github.com/OrchardCMS/OrchardCore/assets/19938236/c91531ba-2fd1-43e9-a014-dcddd7124a8f)

**Comment by jtkech** [comment]
When you are in Development (there is an env variable) and/or in debug moder under Visual Studio, the GC may not work as expected, I already saw this kind of memory increase. Try to build in Release (maybe not necessary), at least start the app from the command line and then look at the Task Manager.

Then it may depend on your machine to decide when you are under memory pressure or not.

**Comment by sebastienros** [comment]
@jtkech is right, the debugger will keep some variables "rooted" so you can inspect them and hence prevent them from being collected.

Use a tool like PerfView or dotnetMemory profiler to take snapshots of the memory from a release run that you can then analyze in these tools.

**Comment by wAsnk** [comment]
I could repro the same result with release mode and `"ASPNETCORE_ENVIRONMENT": "Production"`

**Comment by ShaneCourtrille** [comment]
@wAsnk Can you repro in a deployed environment? You can use dotnet dump collect to get a memory dump and then look at it locally in VS.

**Comment by jtkech** [comment]
@wAsnk 

When I will have time I will re-install PerfView and look at it more in depth, for now my Visual Studio doesn't allow me to do 2 concecutive memory snapshots.

but in the meantime I could repro by just looking at the memory usage, when doing a lot of tenant release the GC works as expected, in my case it stabilizes around 700 Mb (but in debug mode). But yes after having enabled the `ReCaptcha.Users` the memory increases much more with the same test.

I will look at this asap.

Did you see the same behavior but by enabling another feature?

---

@ShaneCourtrille and others

About the `"mvc.1.0.view"` strings, under the debugger by using the `Make object ID` command I could see that all our `RazorViewCompiledItem` that we add (e.g. one for each Liquid file) are all using the same string instance.

But yes, all `DefaultRazorCompiledItem` compiled at build time are all using a different instance, but I'm afraid that here we can't do anything.

**Comment by jtkech** [comment]
@wAsnk 

`ReCaptchaLoginFilter` is activated on each request and resolves transitively `ReCaptchaClient` for which a typed HttpClient has been registered. So on each action request a typed HttpClient is built.

Resolving `ReCaptchaClient` only on demand in `ReCaptchaService` fixed the issue.

Note: Maybe still a memory problem with typed HttpClient but less critical if only built as needed.

**Comment by wAsnk** [comment]
> Did you see the same behavior but by enabling another feature?

No, I couldn't repro with every other feature on.

> Resolving ReCaptchaClient only on demand in ReCaptchaService fixed the issue.

Sounds good!

**Comment by BenedekFarkas** [comment]
> Note: Maybe still a memory problem with typed HttpClient but less critical if only built as needed.

This reminds me of this PR: https://github.com/dotnet/corefx/pull/19082, might have some useful insights there.

**Comment by sebastienros** [comment]
We need to use `IHttpClientFactory` in `ReCaptchaClient` instead of injecting the http client. And we might be able to make `ReCaptchaClient` a singleton (and ReacptchaService by resolving _robotDetectors lazily?)

**Comment by jtkech** [comment]
`IHttpClientFactory` is already used indirectly, a **typed** `HttpClient` is registered meaning that each time `ReCaptchaClient` is resolved an `HttpClient` is build using the `IHttpClientFactory`.

https://github.com/dotnet/runtime/blob/1ff7fbb4c61f08d36afe0e113a676a0fb332ffbc/src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs#L545C17-L545C35

    services.AddHttpClient<ReCaptchaClient>()
        .AddTransientHttpErrorPolicy(policy =>
            policy.WaitAndRetryAsync(3, attempt => TimeSpan.FromSeconds(0.5 * attempt)));

So the problem is because of an mvc action filter that resolves on each request `ReCaptchaService`, then `ReCaptchaClient` which triggers an `HttpClent` building by the `IHttpClientFactory`.

To fix the issue I only needed in `ReCaptchaService` to resolve lazily `ReCaptchaClient` as needed, I already did the change locally, but didn't commit it yet because maybe some other updates may be worth to do, for example maybe better to make `ReCaptchaService` a scoped service, I will check.

For now the only thing I needed to update is only one line, I will commit it to show what I did.

**Comment by jtkech** [comment]
Here what I did for now #14335

**Comment by sebastienros** [comment]
So I read the documentation for `IHttpClientFactory` and it says it's good for short live HttpClient instances. It means they should be disposed after each usage. That would mean that your suggestion might be better then, as long as we only instantiate it when we need the request, and it's disposed.

But using a single static instance share by all tenants would also work if we don't set the base url and pass it on the Post call instead.

**Comment by jtkech** [comment]
Looking at https://learn.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests

Where they talk about disposing or not `HttpClient` and then recommend to use `IHttpClientFactory` for example because of this.

> However, the issue isn't really with HttpClient per se, but with the [default constructor for HttpClient](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.-ctor#system-net-http-httpclient-ctor), because it creates a new concrete instance of [HttpMessageHandler](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpmessagehandler), which is the one that has sockets exhaustion and DNS changes issues mentioned above.

And then for example this

> Manage the lifetime of [HttpMessageHandler](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpmessagehandler) to avoid the mentioned problems/issues that can occur when managing HttpClient lifetimes yourself.

About sharing things accross tenants the problem is that they use non thread safe collections, and we can configure different tenants at the same time, that's why we isolate httpclients registrations per tenants and from the host.

**Comment by jtkech** [comment]
Finally in #14335 I kept the typed client pattern

> For the sake of brevity, this guidance shows the most structured way to use IHttpClientFactory, which is to use Typed Clients (Service Agent pattern).

**Comment by jtkech** [comment]
@rjpowers10 

I tried your simpler test

> I next tried a much simpler example. This next example is a plain old OC site (latest from the main branch) using the blog recipe. That's it. ASPNETCORE_ENVIRONMENT is set to "Development". I then wrote a small PowerShell script to ping the site over and over.

Here I don't think that's an issue, I got the same under the Diagnostic Tools of Visual Studio, the memory increases until it stabilizes around 766 Mb in my case (issexpress.exe 674 Mb), around 550 Mb in your screen shot.

I will think about your original issue where you saw w3wp using over 8.5 GB.

**Comment by rjpowers10** [comment]
> @rjpowers10
> 
> I tried your simpler test
> 
> > I next tried a much simpler example. This next example is a plain old OC site (latest from the main branch) using the blog recipe. That's it. ASPNETCORE_ENVIRONMENT is set to "Development". I then wrote a small PowerShell script to ping the site over and over.
> 
> Here I don't think that's an issue, I got the same under the Diagnostic Tools of Visual Studio, the memory increases until it stabilizes around 766 Mb in my case (issexpress.exe 674 Mb), around 550 Mb in your screen shot.
> 
> I will think about your original issue where you saw w3wp using over 8.5 GB.

It's very possible it is something I'm doing in my code. I do use `HttpClient` a lot to communicate with a REST API, but I'm always going through `IHttpClientFactory` to get the client, so I _think_ that's OK.

**Comment by jtkech** [comment]
Ah okay, thanks for the info.

Yes that's good to use `IHttpClientFactory` but same issue if you create an http client for example on each request, even if you don't need it.

**Comment by jtkech** [comment]
Are you using HttpClient handlers?

**Comment by MikeAlhayek** [comment]
@sebastienros I am still trying to understand this as it does not make sense to me.

![image](https://github.com/OrchardCMS/OrchardCore/assets/24724371/c3e409a8-1b4d-4488-93e5-158f55007c9d)

string interning is managed even across multiple dlls. So even when every view generate a new dll, it should not matter. Here is what I did, created two separate projects in the same solutions 

In Project 1 I added the following 2 classes
```
public class ModuleAClassA
{
    public string First { get; set; } = "First";
}

public class ModuleAClassB
{
    public string First { get; set; } = "First";
}
```

In project 2, I added the following 2 classes

```
public class ModuleBClassA
{
    public string First { get; set; } = "First";
}

public class ModuleBClassB
{
    public string First { get; set; } = "First";
}
```
Then from project 1 I ran the following

```
Console.WriteLine(object.ReferenceEquals(aa.First, ab.First));
Console.WriteLine(object.ReferenceEquals(ba.First, bb.First));
Console.WriteLine(object.ReferenceEquals(ba.First, ab.First));
Console.WriteLine(object.ReferenceEquals(aa.First, bb.First));
Console.WriteLine(object.ReferenceEquals(aa.First, "First"));
```

All 5 lines printed `True` which proves to me that string interning is managed across all dlls. Unless the environment makes a difference like running on Windows vs running on Lunix. Just thought I would share this just in case you have more ideas that you like to share.

**Comment by jtkech** [comment]
Maybe I found a way to remove the `"mvc.1.0.view"` duplicates.

Under the debugger you can see that now they refer to the same instance ID.

![Razor Strings Debug](https://github.com/OrchardCMS/OrchardCore/assets/8586360/fc62bc8c-d352-4e70-ae4e-1d948f3ba5ef)

And no longer in the `Duplicate Strings`.

![Razor Strings Memory](https://github.com/OrchardCMS/OrchardCore/assets/8586360/1f793fa4-3d36-488a-9d29-82e0085b168e)

**Comment by MikeAlhayek** [comment]
Did you do that by using a constant for the string?

**Comment by jtkech** [comment]
No for now I only used `"mvc.1.0.view"` litteral strings that normally are interned in only one.

We only add few compiled items and one per liquid file, most of compiled items are built by aspnetcore by using data loaded from the assemblies.

What I did for all descriptors is to change its item by a new one using this literal for its `Kind` property.

Maybe for perf it's better to use a const, there is already one defined in aspnetcore, so that it doesn't have to lookup in the intern pool.

**Comment by rjpowers10** [comment]
> Ah okay, thanks for the info.
> 
> Yes that's good to use `IHttpClientFactory` but same issue if you create an http client for example on each request, even if you don't need it.

I understand that creating an `HttpClient` when I don't need it is wasteful, but the GC should still be able to clean this up, right? I think my real problem isn't that memory is high but that memory is unable to be reclaimed.

**Comment by sebastienros** [comment]
> but that memory is unable to be reclaimed

If the GC can't collect it it's because it's root, this would be a bug in Orchard or its dependencies, and we need to find it.

**Comment by rjpowers10** [comment]
> > but that memory is unable to be reclaimed
> 
> If the GC can't collect it it's because it's rooted, this would be a bug in Orchard or its dependencies, and we need to find it.

I'm no expert in this area but I'll try to share some things. These may just be red herrings.

Looking at the memory diff, whether sorted by count diff or size diff, `string` is the top of the list.
![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/c76462df-3a9a-4133-b0a2-315d2224dccc)

May or may not be relevant: many of these strings are sitting in the large object heap.
![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/fe461766-b5c2-4857-816c-82c89279571d)

Here's where I start getting confused. Just sampling one of the string instances, the path to root is super long. Maybe this is fine, just seemed odd to me. Maybe this can point someone in the right direction.

1 of 5
![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/598d3ea6-0c3b-4e15-8969-6d049b464978)

2 of 5
![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/925fa96e-e008-4ffa-8595-7b6200a1cae3)

3 of 5
![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/f7fc313f-8890-4c70-bd3c-5c7ec148f99d)

4 of 5
![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/e6dbddc5-518e-4f62-adeb-38934bceb555)

5 of 5
![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/25ee97a4-51d7-4d3e-a1ac-7838727ac36b)

**Comment by jtkech** [comment]
@rjpowers10 

> So I can understand that making that lazy could help, but at first glance that doesn't seem to explain why w3wp is taking over 8GB of memory.

Same answer as @sebastienros but now we may have something more specific to search on.

Maybe around `HttpClientFactory` or `HttpMessageHandlerBuilder`, I will look at it but need to install tools, for now my diagnostic tool often gets stuck, even if I use the last preview of Visual Studio ;)

**Comment by rjpowers10** [comment]
> @rjpowers10
> 
> > So I can understand that making that lazy could help, but at first glance that doesn't seem to explain why w3wp is taking over 8GB of memory.
> 
> Same answer as @sebastienros but now we may have something more specific to search on.
> 
> Maybe around `HttpClientFactory` or `HttpMessageHandlerBuilder`, I will look at it but need to install tools, for now my diagnostic tool often gets stuck, even if I use the last preview of Visual Studio ;)

`IHttpClientFactory` does seem a little suspicious to me. Here's an int array. I'm a little surprised to see it rooted by `DefaultHttpClientFactory`.

![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/aec3c038-26a8-4920-bfd6-77fdb38069ca)

**Comment by jtkech** [comment]
https://github.com/dotnet/runtime/blob/048978abe4cd48222547f3fb8062667cf357904b/src/libraries/Microsoft.Extensions.Http/src/DefaultHttpClientFactory.cs#L21C1-L22C1

**Comment by jtkech** [comment]
Okay, maybe I found something but not so obvious to re-use PerfView so not fully sure.

@rjpowers10 First it seems that it works as expected if we don't release the shell, to repro I need to release the shell/tenant (so rebuilt on a next request) many times, so more like what @wAsnk did.

In your UI Tests are you often releasing the tenant? For example it happens when saving Admin Settings.

---

That said looks like the `DefaultHttpClientFactory` reference the service provider of the current container (our shell context) and has a static `TimerCallback` variable with a delegate that references itself.

No problem in a regular app because `DefaultHttpClientFactory` is a singleton and the host container is not intended to be rebuilt, but in our case it is a tenant singleton where the container can be rebuilt.

So I did a diff where we can see that.

![HttpClientFactory](https://github.com/OrchardCMS/OrchardCore/assets/8586360/923b27c9-8996-41c0-85de-8ae6e037b4e1)

I will investigate more and let you know.

**Comment by rjpowers10** [comment]
> Okay, maybe I found something but not so obvious to re-use PerfView so not fully sure.
> 
> @rjpowers10 First it seems that it works as expected if we don't release the shell, to repro I need to release the shell/tenant (so rebuilt on a next request) many times, so more like what @wAsnk did.
> 
> In your UI Tests are you often releasing the tenant? For example it happens when saving Admin Settings.
> 
> That said looks like the `DefaultHttpClientFactory` reference the service provider of the current container (our shell context) and has a static `TimerCallback` variable with a delegate that references itself.
> 
> No problem in a regular app because `DefaultHttpClientFactory` is a singleton and the host container is not intended to be rebuilt, but in our case it is a tenant singleton where the container can be rebuilt.
> 
> So I did a diff where we can see that.
> 
> ![HttpClientFactory](https://user-images.githubusercontent.com/8586360/268408313-923b27c9-8996-41c0-85de-8ae6e037b4e1.png)
> 
> I will investigate more and let you know.

Nice find! My UI tests are often making changes to admin settings and enabling / disabling features so we can test different things, so yes, we are indirectly releasing the tenant many times throughout the test run.

**Comment by jtkech** [comment]
Okay cool so maybe we found the source of the problem.

We may need to override the `IHttpClientFactory`, maybe not so easy because of internal classes, will see.

**Comment by jtkech** [comment]
@rjpowers10 for info

Okay I could override `DefaultHttpClientFactory`, made it `IDisposable` to nullify some fields, for example `_services`, for now it fixes the issue if we set the options `.SuppressHandlerScope` to true, otherwise still some memory increase (but much less).

I also started to prevent the accumulation of timer callbacks that cleanup the handlers after their lifetime, but the lifetime can be higher than the period we use to release the tenant, but I made some progress.

**Comment by jtkech** [comment]
Applying the above in #14348 seems to fix the issue.

So if you can redo your tests by using this branch, when you will have time.

---

I noticed the same kind of memory increase by doing may times a tenant Reload that not only Releases the tenant but also Reloads and then Rebuilds the Configuration. So we may need to do the same kind of things for the tenant reloading (not only releasing). I will look at it when I will have time.

**Comment by ShaneCourtrille** [comment]
We definitely have a lot of tenant release/reloads over time.

**Comment by rjpowers10** [comment]
@jtkech Huge improvement! I copied your `TenantHttpClientFactory` PR into my code an ran all our tests. The run took about 55min and at the end w3wp took up just under 3GB. Maybe there is some more work to do (maybe OC, maybe my code) but this is great!

I did not test with the recent ReCaptcha or "mvc.1.0.view" PRs so that might make things even better.

**Comment by rjpowers10** [comment]
@ShaneCourtrille I don't think it matters anymore since a recent PR addressed this, but just for completeness I realized I was looking at my memory dump wrong. I actually have over 260,000 instances of "mvc.1.0.view".

**Comment by jtkech** [comment]
@rjpowers10 

Good to know, thanks.

- Yes, `Releasing` a tenant was not a problem in itself, the first issue was related to `HttpClient`.

- But `Reloading` a tenant where we rebuild the `Configuration` stack still has memory leaks. Normally we only reload a tenant on setup or when enabling / disabling features, not when updating settings, so if you have time you could retry your tests with all the features already enabled, if possible.

    I will work on this part related to tenant `Reloading` asap, maybe a separate PR.

- About razor strings I may ave some ideas to reduce the number but I don't think they are rooted.

**Comment by jtkech** [comment]
Okay, I may have found the source of the 2nd problem related to tenant Reloading where we rebuild the Configuration stack using some providers inheriting from `FileConfigurationProvider`.

See https://github.com/dotnet/runtime/issues/86146 and https://github.com/dotnet/runtime/pull/86455

**Comment by jtkech** [comment]
Okay fixed locally but need to be tweaked, I will update the PR soon.

**Comment by ShaneCourtrille** [comment]
@rjpowers10 I'm still not done looking but my initial checks look good for the mvc.1.0.view fix and in our case that saves us 500MB of memory alone.

**Comment by sebastienros** [comment]
![ThrillerMichaelJacksonGIF](https://github.com/OrchardCMS/OrchardCore/assets/1165805/fae5a43d-2c01-4cb1-a005-ca9375041d62)

**Comment by jtkech** [comment]
@sebastienros Not related ;)

@rjpowers10 @ShaneCourtrille 

I made good progress in #14348 for the Reloading part, not fully tweaked but can be already tested.

Note: I was wrong, the tenant is not reloaded on Features change, it is only released and then rebuilt based on the updated shell descriptor. The tenant is reloaded from the Admin Tenants UI or by code.

But maybe you are reloading tenants in your tests, anyway worth to re-try as I did other changes.

Didn't have time to focus on the razor strings, I will do, didn't see yet the meeting where it was discussed.

**Comment by MikeAlhayek** [comment]
@jtkech could the actual issue with HttpClient is the usage by not disposing it? Meaning doing this

```C#
var client = _httpClientFactory.CreateClient();

// never dispose the client
```

instead of 

```C#
using var client = _httpClientFactory.CreateClient();

```

or

```
var client = _httpClientFactory.CreateClient();

client.Dispose();
```

**Comment by rjpowers10** [comment]
@MikeAlhayek 

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-7.0#httpclient-and-lifetime-management

![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/1bd0dcaa-1869-4a9c-9598-f16342d3c2fb)

**Comment by jtkech** [comment]
@rjpowers10 @ShaneCourtrille and others.

Here a little summary of what has been done done in #14348, if you can give it a try.

- So in #14348 we override the `IHttpClientFactory`, mainly to make it `IDisposable` so that when releasing / reloading a Tenant we release clients Handlers and cleanup Timers.

- On Tenant Reloading, we now also Release the 2 inner `IConfiguration` of `ShellSettings`.

- Finally we now always use our static shared `SharedViewCompilerProvider` (see below).

About razor view compiled items, we already have a static shared compiler provider but that we only use if razor runtime compilation is enabled, now in #14348 we always register it.

            // Shares across tenants a static compiler even if there is no runtime compilation
            // because the compiler still uses its internal cache to retrieve compiled items.
            services.AddSingleton<IViewCompilerProvider, SharedViewCompilerProvider>();

We still load the `CompiledDescriptor` from assemblies but not to be hold by another compiler instance, the cache of the same static compiler is populated once.

Doing so it seems that the GC behavior is better, here after many Tenant Release / Reload.

 ![Memory Leaks](https://github.com/OrchardCMS/OrchardCore/assets/8586360/ef252092-6ea0-4986-8dfe-8ff107e045bf)

![Memory GC](https://github.com/OrchardCMS/OrchardCore/assets/8586360/c9d6c2a5-8e6a-430c-a44b-cd17b544d722)

**Comment by rjpowers10** [comment]
@jtkech Awesome! I will test what I can, but I can only really test the stuff that can be overridden in DI. I've just been copy-pasting your PR changes into my code, since I will need these fixes before the next OC release.

**Comment by wAsnk** [comment]
@jtkech I tested it in our own solution with local NuGet files and it looks really good. I will do further testing in the upcoming days.

**Comment by jtkech** [comment]
@rjpowers10 No problem, no pressure, when you will have time.

@wAsnk Okay cool, thanks for testing, I cross my fingers.

**Comment by jtkech** [comment]
@wAsnk 

Just for info in #14348 `Recaptcha.Users` has not been changed, but still better to create `HttpClient` on demand as done in #14335. So you may see more memory usage but because we override `DefaultHttpClientFactory` normally, at an higher value, but it should stabilize.

**Comment by jtkech** [comment]
@rjpowers10 @ShaneCourtrille @MikeAlhayek @wAsnk @sebastienros

I saw one of the last meeting related to string allocations, interesting. For info by using instance ids under the debugger we can see that compiled items loaded from assemblies hold different instances of the `"mvc.1.0.view"` string.

![Razor string duplicates](https://github.com/OrchardCMS/OrchardCore/assets/8586360/1833ddb5-a616-47ca-aa73-f15766cf334e)

In #14348 we now override how compiled items are loaded and we can see that the same instance is used.

![Razor string no duplicate](https://github.com/OrchardCMS/OrchardCore/assets/8586360/1ab13ccc-fa47-409c-ae75-db29d7f38f29)

Also I could reduce the number of times the compiled items are loaded, before 3 times per tenant building, for example the view descriptors are populated on demand by models providers (one from aspnetcore, one from orchardcore), now only once by caching the descriptors built in a given shell scope.

We can still have duplicate strings but they are no longer rooted and can be reclaimed, the only one holding these strings being our static compiler that we now always use, and now it only references one instance of `"mvc.1.0.view"`.

**Comment by wAsnk** [comment]
I did some testing with Azure Web App.

TLDR: LGTM

I did some mild testing yesterday, no real surprises the memory usage went up, then stayed with some sawtooth shapes:
![image](https://github.com/OrchardCMS/OrchardCore/assets/19938236/a0054071-5cb2-41c9-a5a9-1ead82ac3a5b)

But there were no real cleanups like I would expect to GC down to the initial 80-120MB of usage. So I did some heavy testing today to really kick in GC on heavy memory usage. Around 1.4GB GC pressure was really high and did its major cleaning, the memory usage went back to 90MB, just like it happened locally for the same solution around 1GB.
![image](https://github.com/OrchardCMS/OrchardCore/assets/19938236/b9932144-ab56-4d87-8524-8fecd42d944b)

**Comment by rjpowers10** [comment]
Sorry, I've been meaning to run some more tests on my solution but I've been swamped this week. The PR has also grown considerably, to the point where it would be easier if I had NuGet packages. Maybe I will pull the branch and make some local packages to test with.

**Comment by jtkech** [comment]
@wAsnk Thanks a lot for your tests.

Yes, I've made other changes, for example to not call `CreateChildContainer()` twice to build a shell, and as said, in #14348 I haven't intentionally updated the "RecaptchaClient on demand" yet, so perhaps less memory increase with these changes.

But I think we are okay the main goal being to keep the memory reclaimable by the GC.

What's your own conclusion?

@rjpowers10 No problem, only when you will have time.

**Comment by wAsnk** [comment]
> But I think we are okay the main goal being to keep the memory reclaimable by the GC.

Yeah, I agree.

**Comment by jtkech** [comment]
Okay cool then, thank's again for your tests.

**Comment by wAsnk** [comment]
@jtkech When [Database Shells Configuration Provider](https://docs.orchardcore.net/en/latest/docs/reference/core/Shells/#database-shells-configuration-provider) is used there is an exception on startup:

```
System.InvalidOperationException: StreamConfigurationProviders cannot be loaded more than once.
   at Microsoft.Extensions.Configuration.StreamConfigurationProvider.Load()
   at Microsoft.Extensions.Configuration.ConfigurationRoot.Reload()
   at OrchardCore.Environment.Shell.ShellSettingsManager.LoadSettingsAsync(String tenant)
   at OrchardCore.Environment.Shell.Distributed.DistributedShellHostedService.LoadingAsync()
   at OrchardCore.Environment.Shell.ShellHost.PreCreateAndRegisterShellsAsync()
   at OrchardCore.Environment.Shell.ShellHost.InitializeAsync()
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

**Comment by jtkech** [comment]
Okay, fixed by #14499

**Comment by rjpowers10** [comment]
For posterity, I updated my app to Orchard Core 1.8, the first release containing the fixes from #14348.

In this graph you can see a big improvement in memory consumption between OC 1.7 and 1.8.

![image](https://github.com/OrchardCMS/OrchardCore/assets/68876423/e9e5003c-c9bc-4dd0-86f6-cea7bcda3b38)

**Comment by hyzx86** [comment]
I created a background task to collect garbage every minute
```
    [BackgroundTask(Schedule = "* * * * *", Description = "Release Memory. GC Collect")]
    public class JintBackgroundTask : IBackgroundTask
    {
        public Task DoWorkAsync(IServiceProvider serviceProvider, CancellationToken cancellationToken)
        {
            GC.Collect();
            return Task.CompletedTask;
        }
    }
```

**Comment by Piedone** [comment]
This seems unnecessary and potentially degrading for the overall performance, since [it's blocking](https://learn.microsoft.com/en-us/dotnet/api/system.gc.collect) and while may reclaim memory, uses CPU. Outside of debugging and otherwise very special scenarios, you shouldn't (have to) run the GC manually.

Note that on servers the point is not to keep memory usage low, but to serve the most users with good performance. This may mean keeping memory usage high, but steady. This in itself is not a problem.

**Comment by hyzx86** [comment]
This is a test from a long time ago, I tried to disable it and see what happens with the automatic memory reclaim

**Comment by rjpowers10** [comment]
Agreed with @Piedone that we shouldn't be calling `GC.Collect()` manually, but for completeness I'll also add that it wouldn't have resolved this particular issue anyway, since the memory was still rooted and therefore couldn't be reclaimed.
